test/PandoraInterface.cxx

```diff
@@ -336,14 +368,462 @@ void ProcessEvents(const Parameters &parameters, const Pandora *const pPrimaryPa
+        currentVoxelList.push_back(voxel);
+
+        // Update ray starting position using intersection path difference
+        const pandora::CartesianVector newStart = ray.getPoint(dL);
+        ray.updateOrigin(newStart);
+        loop++;
+    }
+
+    //std::cout << "Returning vector of " << currentVoxelList.size() << " voxels for G4HitSegment\n" << std::endl;
+
+    return currentVoxelList;
+}
+
+bool LArBox::intersect(const LArRay &ray, double &t0, double &t1) const
+{
+    // Brian Smits ray-box algorithm with improvements from Amy Williams et al
+    double x(ray.m_origin.GetX());
+    double y(ray.m_origin.GetY());
+
+    double invDirX(ray.m_invDir.GetX());
+    double invDirY(ray.m_invDir.GetY());
+
+    double tMin(0.0), tMax(0.0), tyMin(0.0), tyMax(0.0), tzMin(0.0), tzMax(0.0);
+
+    if (ray.m_sign[0] == 0)
+    {
+        tMin = (m_bottom.GetX() - x) * invDirX;
+        tMax = (m_top.GetX() - x) * invDirX;
+    }
+    else
+    {
+        tMin = (m_top.GetX() - x) * invDirX;
+        tMax = (m_bottom.GetX() - x) * invDirX;
+    }
+
+    if (ray.m_sign[1] == 0)
+    {
+        tyMin = (m_bottom.GetY() - y) * invDirY;
+        tyMax = (m_top.GetY() - y) * invDirY;
+    }
+    else
+    {
+        tyMin = (m_top.GetY() - y) * invDirY;
+        tyMax = (m_bottom.GetY() - y) * invDirY;
+    }
+
+    if ((tMin > tyMax) || (tyMin > tMax))
+    {
+        return false;
+    }
+
+    if (tyMin > tMin)
+    {
+        tMin = tyMin;
+    }
+
+    if (tyMax < tMax)
+    {
+        tMax = tyMax;
+    }
+
+    double z(ray.m_origin.GetZ());
+    double invDirZ(ray.m_invDir.GetZ());
+    if (ray.m_sign[2] == 0)
+    {
+        tzMin = (m_bottom.GetZ() - z) * invDirZ;
+        tzMax = (m_top.GetZ() - z) * invDirZ;
+    }
+    else
+    {
+        tzMin = (m_top.GetZ() - z) * invDirZ;
+        tzMax = (m_bottom.GetZ() - z) * invDirZ;
+    }
+
+    if ((tMin > tzMax) || (tzMin > tMax))
+    {
+        return false;
+    }
+
+    if (tzMin > tMin)
+    {
+        tMin = tzMin;
+    }
+
+    if (tzMax < tMax)
+    {
+        tMax = tzMax;
+    }
+
+    // First and second intersection path lengths
+    t0 = tMin;
+    t1 = tMax;
+    return true;
+}
+
+bool LArBox::inside(const pandora::CartesianVector &point) const
+{
+    bool result(false);
+
+    float x = point.GetX();
+    float y = point.GetY();
+    float z = point.GetZ();
+
+    if (x > m_bottom.GetX() && x < m_top.GetX() && y > m_bottom.GetY() && y < m_top.GetY() && z > m_bottom.GetZ() && z < m_top.GetZ())
+    {
+        result = true;
+    }
+
+    return result;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+std::vector<LArVoxel> MergeSameVoxels(const std::vector<LArVoxel> &voxelList)
+{
+
+    std::cout << "Merging voxels with the same IDs" << std::endl;
+    std::vector<LArVoxel> mergedVoxels;
+
+    const int nVoxels = voxelList.size();
+    std::vector<int> processed(nVoxels, 0);
```

Comment:
strictly std::vector<bool>?

Reply:
Changed to use booleans. I prefer to keep this as a vector to preserve the (initial) voxel ordering. The loop over voxel pairs (i and j) gets progressively faster, since all earlier voxels will be processed beforehand.